### ISSUE TITLE
How to ensure that messages are processed in order

### ISSUE THREAD

**Issue by suntianji123** [topfreegames/pitaya#179]

**Comment by felipejfc** [comment]
what do you mean? inside the same session, it's expected that the messages will be processed in order because we're using either TCP or WS as acceptor. unless you're trying to implement an UDP acceptor then I don't see why this would be a concern to you

**Comment by edisonwsk** [comment]
Data received from the same player may be processed in different coroutines, resulting in inconsistent processing order and receiving order

**Comment by leohahn** [comment]
> Data received from the same player may be processed in different coroutines
This should not be the case, since we process all messages on the same goroutine:
https://github.com/topfreegames/pitaya/blob/177cdf3913e6bd7fa5d0f0ac01dca4c7400a3985/service/handler.go#L183

Are you getting a case where this is happening? Or is it just a speculation?

**Comment by edisonwsk** [comment]
> https://github.com/topfreegames/pitaya/blob/177cdf3913e6bd7fa5d0f0ac01dca4c7400a3985/service/handler.go#L119
Here is the concurrent execution！

**Comment by leohahn** [comment]
Oh, so you're talking about `clients` as other servers right? Because indeed, messages are not ordered between servers communicating.


However, when I mentioned client, I was talking about the device itself, and not another server.

**Comment by edisonwsk** [comment]
The behavior of the same user should be orderly, so the messages of the same user should be processed in the same coroutine to avoid concurrent processing. The current general processing method is to determine the player's coroutine based on sessionid%threads

**Comment by felipejfc** [comment]
Inside a same server, the messages of the same users are always ordered because they're processed in a single coroutine.

If you have multiple servers, you may end up with unordering indeed, the mechanism that pitaya provides to you so that you avoid this is using a [router](https://github.com/topfreegames/pitaya/blob/master/router/router.go). With it you can ensure that messages from a same client will always end up in the same backend server, and then because of what I said above you will guarantee processing order.

**Comment by leohahn** [comment]
Indeed, what I said before was wrong. The dispatch code also processes messages from the client. Really good find. IMO, I think having only one coroutine for processing player messages is enough. The low message count ensures that one coroutine would not be an issue in terms of performance.

**Comment by leohahn** [comment]
Another potential option would be to have a concept or "ordered packets" from the client. So we would then have two coroutines, one for multiple coroutine processing, and another channel only for messages that need to be processed in order. This is a larger change on the codebase, however, and I'm not sure if this use case is actually needed.

**Comment by edisonwsk** [comment]
There is a solution: each coroutine has an independent processing channel, and then the user only posts the message to the coroutine where sessionid%threads is located for processing. In theory, all users can be equally distributed to different coroutines, but a certain coroutine is not excluded. Process blocking or abnormal exit causes all users assigned to the coroutine to be blocked, which should require more safeguards. Do you have a better way to deal with it?

**Comment by leohahn** [comment]
This seems to be a good solution overall. I don't see many issues here apart from the potential issue of client starving that you mentioned. I assume the consuming threads already have a recover from panics mechanism for handling these issues, but I'm not sure now.

My only concern with this solution is that the distribution of clients to channels could be an issue if we had clients with different connection times. For example, assuming all clients send an equal amount of messages and stay connected for the same amount of time, we would not have an issue with "hot" coroutines. However, we might have a difference of message frequency and connection time from clients, which might end up creating hot coroutines, and therefore slow down consumption for some clients.

**Comment by edisonwsk** [comment]
Yes, this is what we are worried about that some coroutines will be heavily loaded after a period of time, and some coroutines will be idle. But it is also possible that the distribution is not too concentrated in terms of probability because of the even distribution. At present, we do not have a better solution for the time being, but we can evaluate its pros and cons based on the test data in the future. Of course, it would be better if you have a better plan. Thanks for the answer!

**Comment by chgz** [comment]
@edisonwsk  i have implemented this solution, but I gave up in the end.
because it is imperfect. 
it can only take effect in limited scenarios:  all rpc use sticky routing or target server is standalone.

eg.  when a player client send requests msg1,msg2... to backend server (serviceTypeA, suppose there are 3 nodes)  , msg1 arrived node1, msg2 arrived node2...
it cannot guarantee the processing order of requests.

**Comment by leohahn** [comment]
Thinking about it in a simpler way, did you consider just guaranteeing the ordering on client-side? For example, you can wait for the response on the client (libpitaya) before sending the next request.

**Comment by chgz** [comment]
it is similar to the BIO model.
in some scenarios, it’s not enough

in fact, not all client messages require sequentiality. In my projects, the most common is the combat messages, but their goal is unique, so this is not a troublesome problem for me.

**Comment by edisonwsk** [comment]
@chgz 
Sorry, I did not understand what you said. I think that if a player's message is fixed in a coroutine for execution, then according to the nature of rpc's blocking and waiting to return, the execution of the message must be orderly.
> https://github.com/topfreegames/pitaya/blob/177cdf3913e6bd7fa5d0f0ac01dca4c7400a3985/service/handler.go#L132
In some cases, player actions need to be orderly. For example, action A is a precondition for action B. The player quickly presses A and then B. In the case of a bad network, the two data of AB may be merged. , The server may continuously push two messages into the queue for execution. If they are processed in two coroutines, then it is possible that B has started to execute, and A has not been executed yet, which does not match the player's expected action.

**Comment by chgz** [comment]
Sent by the same dispatcher, they are in order. 
But receiver is not necessarily the same. 
So，in the case of multiple nodes and long call chains,  we can't expect the strong order.

**Comment by edisonwsk** [comment]
We don’t need to care about the order of subsequent actions. We only need to consider that the actions of a single player are in order. It is impossible to execute concurrently if all nodes are in order.

**Comment by tutumagi** [comment]
My project is working with `pitaya` for a while. I think `pitaya` is more fit for Hall-Room Game. Not fit for MMO Game. MMO Game state synchronization is a big problem. And it need process single player message in order I think. So I am looking for `protoactor.go`. 
PS. I love the route/metrics/cluster/span implement in `pitaya`.

**Comment by chgz** [comment]
I agree with your opinion. @tutumagi 

But, Go’s concurrency model relies on CSP. 
If you want to use the actor model, the complexity will increase a lot.
I tend to replace the random routing algorithm with consistent hashring.

**Comment by bruce1125** [comment]
> od find. IMO, I

To hash the messages, pitaya need to know which "actor" is bound with the message not only the client messages but also remote messages.
